File: fsw/bus/exchange.c
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <hal/clock.h>
#include <hal/debug.h>
#include <hal/init.h>
#include <hal/thread.h>
#include <bus/exchange.h>

enum {
    EXCHANGE_REPLICA_ID = 0,
};

//#define EXCHANGE_DEBUG

#define debug_printf(lvl, fmt, ...) debugf(lvl, "[%s] " fmt, exc->conf->label, ## __VA_ARGS__)

static void rand_init(void) {
    // this does have to be deterministic for our simulations...
    srand(1552);
}
PROGRAM_INIT(STAGE_RAW, rand_init);

// random interval in the range [1 tick, 5 ticks]
static uint32_t handshake_period_ticks(void) {
    return rand() % 4 + 1;
}

static void exchange_instance_configure(exchange_instance_t *exc, fw_exchange_t *conf, uint32_t countdown_timeout) {
    assert(exc != NULL && conf != NULL);

    exc->conf = conf;

    exc->exc_state    = FW_EXC_CONNECTING;
    exc->recv_state   = FW_RECV_LISTENING;

    exc->countdown_timeout = countdown_timeout;

    exc->send_handshake_id = 0;
    exc->recv_handshake_id = 0;

    exc->send_primary_handshake   = false;
    exc->send_secondary_handshake = false;

    exc->fcts_sent   = 0;
    exc->fcts_rcvd   = 0;
    exc->pkts_sent   = 0;
    exc->pkts_rcvd   = 0;
    exc->resend_fcts = false;
    exc->resend_pkts = false;

    exc->read_offset = 0;
    exc->read_timestamp = 0;
    exc->write_needs_error = false;
}

static void exchange_instance_reset(exchange_instance_t *exc) {
    assert(exc != NULL);
    exchange_instance_configure(exc, exc->conf, exc->countdown_timeout);
}

static void exchange_instance_check_timers(exchange_instance_t *exc) {
    assert(exc != NULL);

    // check timers now
    if (exc->countdown_timeout > 0) {
        exc->countdown_timeout -= 1;
    }
    if (exc->countdown_timeout == 0) {
        if (exc->exc_state == FW_EXC_OPERATING) {
            // send heartbeats
            exc->resend_fcts = true;
            exc->resend_pkts = true;

            exc->countdown_timeout = handshake_period_ticks();
        } else if (exc->exc_state == FW_EXC_HANDSHAKING || exc->exc_state == FW_EXC_CONNECTING) {
            // send a fresh handshake
            exc->send_primary_handshake = true;

            exc->countdown_timeout = handshake_period_ticks();
            debug_printf(DEBUG, "Next handshake scheduled for %u ticks in the future", exc->countdown_timeout);
        }
    }
}

static void exchange_instance_check_invariants(exchange_instance_t *exc) {
    assert(exc != NULL);

    // check invariants
    assert(exc->exc_state >= FW_EXC_CONNECTING && exc->exc_state <= FW_EXC_OPERATING);
    assertf(exc->pkts_sent <= exc->fcts_rcvd && exc->fcts_rcvd <= exc->pkts_sent + MAX_OUTSTANDING_TOKENS,
            "pkts_sent = %u, fcts_rcvd = %u", exc->pkts_sent, exc->fcts_rcvd);
}

static void exchange_recv_ctrl_char_while_connecting(exchange_instance_t *exc, fw_ctrl_t symbol, uint32_t param) {
    assert(exc != NULL);

    // error condition: if ANYTHING is hit except a handshake_1
    if (symbol != FWC_HANDSHAKE_1) {
        // There's no point in being loud about this; if we're seeing it, we're ALREADY in a broken
        // state, and continuing to spew messages about how everything is still broken is not helpful.
        debug_printf(TRACE, "Unexpected %s(0x%08x) instead of HANDSHAKE_1(*); resetting.",
                     fakewire_codec_symbol(symbol), param);
        exchange_instance_reset(exc);
        return;
    }

    // received a primary handshake
    debug_printf(DEBUG, "Received a primary handshake with ID=0x%08x.", param);
    exc->recv_handshake_id = param;
    exc->send_secondary_handshake = true;
}

static void exchange_recv_ctrl_char_while_handshaking(exchange_instance_t *exc, fw_ctrl_t symbol, uint32_t param) {
    assert(exc != NULL);

    // error condition: if ANYTHING is hit except a match handshake_2
    if (symbol != FWC_HANDSHAKE_2 || param != exc->send_handshake_id) {
        debug_printf(WARNING, "Unexpected %s(0x%08x) instead of HANDSHAKE_2(0x%08x); resetting.",
                     fakewire_codec_symbol(symbol), param, exc->send_handshake_id);
        exchange_instance_reset(exc);
        return;
    }

    // received a valid secondary handshake
    debug_printf(DEBUG, "Received secondary handshake with ID=0x%08x; transitioning to operating mode.", param);
    exc->exc_state = FW_EXC_OPERATING;
    exc->send_primary_handshake = false;
    exc->send_secondary_handshake = false;
}

static void exchange_recv_ctrl_char_while_operating(exchange_instance_t *exc, fw_ctrl_t symbol, uint32_t param,
                                                    uint64_t receive_timestamp) {
    assert(exc != NULL);

    // TODO: act on any received FWC_HANDSHAKE_1 requests immediately.
    switch (symbol) {
    case FWC_START_PACKET:
        if (exc->fcts_sent <= exc->pkts_rcvd) {
            debug_printf(WARNING, "Received unauthorized start-of-packet (fcts_sent=%u, pkts_rcvd=%u); resetting.",
                         exc->fcts_sent, exc->pkts_rcvd);
            exchange_instance_reset(exc);
            break;
        }
        assert(exc->recv_state == FW_RECV_LISTENING);

        // should always be allowed, because the number of fcts we send are based on the max flow rate
        assert(duct_send_allowed(exc->conf->read_duct, EXCHANGE_REPLICA_ID));

        // reset receive state and buffer before proceeding
        exc->read_offset = 0;
        // Had to disable this memset because it was too slow to be practical.
        // memset(exc->conf->read_buffer, 0, exc->conf->buffers_length);
        exc->read_timestamp = receive_timestamp;

        exc->recv_state = FW_RECV_RECEIVING;
        exc->pkts_rcvd += 1;
        break;
    case FWC_END_PACKET:
        if (exc->recv_state == FW_RECV_OVERFLOWED) {
            // discard state and get ready for another packet
            exc->recv_state = FW_RECV_LISTENING;
        } else if (exc->recv_state != FW_RECV_RECEIVING) {
            debug_printf(WARNING, "Hit unexpected END_PACKET in receive state %d; resetting.", exc->recv_state);
            exchange_instance_reset(exc);
        } else if (exc->read_offset == 0) {
            debug_printf(WARNING, "Packet of length 0 received; discarding.");
            exc->recv_state = FW_RECV_LISTENING;
        } else {
            // transmit received packet through duct
            duct_send_message(exc->conf->read_duct, EXCHANGE_REPLICA_ID,
                              exc->conf->read_buffer, exc->read_offset, exc->read_timestamp);
            exc->recv_state = FW_RECV_LISTENING;
        }
        break;
    case FWC_ERROR_PACKET:
        if (exc->recv_state != FW_RECV_OVERFLOWED && exc->recv_state != FW_RECV_RECEIVING) {
            debug_printf(WARNING, "Hit unexpected ERROR_PACKET in receive state %d; resetting.", exc->recv_state);
            exchange_instance_reset(exc);
            break;
        }
        // discard state and get ready for another packet
        exc->recv_state = FW_RECV_LISTENING;
        break;
    case FWC_FLOW_CONTROL:
        if (param < exc->fcts_rcvd) {
            // FCT number should never decrease.
            debug_printf(WARNING, "Received abnormally low FCT(%u) when last count was %u; resetting.",
                         param, exc->fcts_rcvd);
            exchange_instance_reset(exc);
        } else if (param > exc->pkts_sent + MAX_OUTSTANDING_TOKENS) {
            // FCT number should never increase more than allowed.
            debug_printf(WARNING, "Received abnormally high FCT(%u) when maximum was %u and last count was %u; "
                         "resetting.", param, exc->pkts_sent + MAX_OUTSTANDING_TOKENS, exc->fcts_rcvd);
            exchange_instance_reset(exc);
        } else {
            // received FCT; may be able to send more packets!
            exc->fcts_rcvd = param;
        }
        break;
    case FWC_KEEP_ALIVE:
        if (exc->pkts_rcvd != param) {
            debug_printf(WARNING, "KAT mismatch: received %u packets, but supposed to have received %u; resetting.",
                         exc->pkts_rcvd, param);
            exchange_instance_reset(exc);
        }
        break;
    default:
        debug_printf(WARNING, "Unexpected %s(0x%08x) during OPERATING mode; resetting.",
                     fakewire_codec_symbol(symbol), param);
        exchange_instance_reset(exc);
        if (symbol == FWC_HANDSHAKE_1) {
            // special case: process received handshakes immediately
            exchange_recv_ctrl_char_while_connecting(exc, symbol, param);
        }
    }
}

static void exchange_instance_receive_ctrl_char(exchange_instance_t *exc, fw_ctrl_t symbol, uint32_t param,
                                                uint64_t receive_timestamp) {
    assert(exc != NULL);
    assert(param == 0 || fakewire_is_parametrized(symbol));

#ifdef EXCHANGE_DEBUG
    debug_printf(TRACE, "Received control character: %s(0x%08x).", fakewire_codec_symbol(symbol), param);
#endif

    switch (exc->exc_state) {
    case FW_EXC_CONNECTING:
        exchange_recv_ctrl_char_while_connecting(exc, symbol, param);
        break;
    case FW_EXC_HANDSHAKING:
        exchange_recv_ctrl_char_while_handshaking(exc, symbol, param);
        break;
    case FW_EXC_OPERATING:
        exchange_recv_ctrl_char_while_operating(exc, symbol, param, receive_timestamp);
        break;
    default:
        assert(false);
    }
}

static void exchange_instance_receive_data_chars(exchange_instance_t *exc, uint8_t *data_in, size_t data_len) {
    assert(exc != NULL);

    if (exc->recv_state == FW_RECV_OVERFLOWED) {
        assert(exc->exc_state == FW_EXC_OPERATING);
        assert(data_in == NULL);
        // discard extraneous bytes and do nothing
#ifdef EXCHANGE_DEBUG
        debug_printf(DEBUG, "Discarded an additional %zu regular data bytes.", data_len);
#endif
    } else if (exc->exc_state != FW_EXC_OPERATING || exc->recv_state != FW_RECV_RECEIVING) {
        assert(data_in == NULL);
        debug_printf(WARNING, "Received at least %zu unexpected data characters during state (exc=%d, recv=%d); "
                     "resetting.", data_len, exc->exc_state, exc->recv_state);
        exchange_instance_reset(exc);
    } else if (exc->read_offset >= exc->conf->buffers_length) {
        assert(data_in == NULL);
        debug_printf(WARNING, "Packet exceeded buffer size %zu (at least %zu + %zu bytes); discarding.",
                     exc->conf->buffers_length, exc->read_offset, data_len);
        exc->recv_state = FW_RECV_OVERFLOWED;
    } else {
        assert(data_in != NULL);
        assert(exc->read_offset + data_len <= exc->conf->buffers_length);
#ifdef EXCHANGE_DEBUG
        debug_printf(TRACE, "Received %zu regular data bytes.", data_len);
#endif
        exc->read_offset += data_len;
        assert(exc->read_offset <= exc->conf->buffers_length);
    }
}

static bool exchange_instance_receive(exchange_instance_t *exc) {
    // discard all data and just tell us the number of bytes
    fw_decoded_ent_t rx_ent = {
        .data_out     = NULL,
        .data_max_len = 0,
    };

    // UNLESS we have somewhere we can put that data, in which case put it there
    if (exc->exc_state == FW_EXC_OPERATING && exc->recv_state == FW_RECV_RECEIVING
            && exc->read_offset < exc->conf->buffers_length) {
        rx_ent.data_out     = exc->conf->read_buffer + exc->read_offset;
        rx_ent.data_max_len = exc->conf->buffers_length - exc->read_offset;
    }

    if (!fakewire_dec_decode(&exc->decoder, &rx_ent)) {
        // no more data to receive right now; wait until next wakeup
        return false;
    }
    // process received control character or data characters
    if (rx_ent.ctrl_out != FWC_NONE) {
        assert(rx_ent.data_actual_len == 0);

        exchange_instance_receive_ctrl_char(exc, rx_ent.ctrl_out, rx_ent.ctrl_param, rx_ent.receive_timestamp);
    } else {
        assert(rx_ent.data_actual_len > 0);

        exchange_instance_receive_data_chars(exc, rx_ent.data_out, rx_ent.data_actual_len);
    }

    return true;
}

static void exchange_instance_check_fcts(exchange_instance_t *exc) {
    assert(exc != NULL);

    uint32_t not_yet_received = duct_max_flow(exc->conf->read_duct);
    if (exc->recv_state != FW_RECV_LISTENING) {
        not_yet_received -= 1;
    }
    if (not_yet_received > MAX_OUTSTANDING_TOKENS) {
        not_yet_received = MAX_OUTSTANDING_TOKENS;
    }
    if (exc->exc_state == FW_EXC_OPERATING && exc->pkts_rcvd + not_yet_received > exc->fcts_sent) {
#ifdef EXCHANGE_DEBUG
        debug_printf(TRACE, "Sending FCT.");
#endif
        exc->fcts_sent = exc->pkts_rcvd + not_yet_received;
        exc->resend_fcts = true;
        exc->resend_pkts = true;

        exc->countdown_timeout = handshake_period_ticks();
    }
}

static void exchange_instance_transmit_tokens(exchange_instance_t *exc) {
    assert(exc != NULL);

    if (exc->resend_fcts && fakewire_enc_encode_ctrl(&exc->encoder, FWC_FLOW_CONTROL, exc->fcts_sent)) {
        assert(exc->exc_state == FW_EXC_OPERATING);
        exc->resend_fcts = false;
#ifdef EXCHANGE_DEBUG
        debug_printf(TRACE, "Transmitted reminder FCT(%u) tokens.", exc->fcts_sent);
#endif
    }

    if (exc->resend_pkts && fakewire_enc_encode_ctrl(&exc->encoder, FWC_KEEP_ALIVE, exc->pkts_sent)) {
        assert(exc->exc_state == FW_EXC_OPERATING);
        exc->resend_pkts = false;
#ifdef EXCHANGE_DEBUG
        debug_printf(TRACE, "Transmitted reminder KAT(%u) tokens.", exc->pkts_sent);
#endif
    }
}

static void exchange_instance_transmit_handshakes(exchange_instance_t *exc) {
    assert(exc != NULL);

    if (exc->send_secondary_handshake
            && fakewire_enc_encode_ctrl(&exc->encoder, FWC_HANDSHAKE_2, exc->recv_handshake_id)) {
        assert(exc->exc_state == FW_EXC_CONNECTING);

        exc->exc_state = FW_EXC_OPERATING;
        exc->send_primary_handshake = false;
        exc->send_secondary_handshake = false;

        debug_printf(DEBUG, "Sent secondary handshake with ID=0x%08x; transitioning to operating mode.",
                     exc->recv_handshake_id);

        exc->countdown_timeout = handshake_period_ticks();
    }

    if (exc->send_primary_handshake) {
        assert(exc->exc_state == FW_EXC_HANDSHAKING || exc->exc_state == FW_EXC_CONNECTING);

        // pick something very likely to be distinct (Go picks msb unset, C picks msb set)
        uint32_t gen_handshake_id = 0x80000000 + (0x7FFFFFFF & (uint32_t) clock_timestamp_monotonic());

        if (fakewire_enc_encode_ctrl(&exc->encoder, FWC_HANDSHAKE_1, gen_handshake_id)) {
            exc->send_handshake_id = gen_handshake_id;

            exc->exc_state = FW_EXC_HANDSHAKING;
            exc->send_primary_handshake = false;
            exc->send_secondary_handshake = false;

            debug_printf(DEBUG, "Sent primary handshake with ID=0x%08x; transitioning to handshaking mode.",
                         gen_handshake_id);
        }
    }
}

static bool exchange_instance_transmit_data(exchange_instance_t *exc, size_t length) {
    assert(exc != NULL);

    if (exc->exc_state != FW_EXC_OPERATING) {
        // can't transmit anything until we're in the operating state. drop packets instead.
        return false;
    }

    if (exc->write_needs_error) {
        // if we weren't able to transmit the whole last packet, then we need to make sure to transmit ERROR_PACKET to
        // make sure the remote end drops it instead of trying to process it.
        if (fakewire_enc_encode_ctrl(&exc->encoder, FWC_ERROR_PACKET, 0)) {
            exc->write_needs_error = false;
        } else {
            debugf(TRACE, "Transmit buffer is full.");
            return false;
        }
    }

    if (exc->pkts_sent >= exc->fcts_rcvd) {
        // no flow control tokens received; can't transmit any packets yet. drop them instead.
        debugf(TRACE, "No more flow control tokens available.");
        return false;
    }
    if (!fakewire_enc_encode_ctrl(&exc->encoder, FWC_START_PACKET, 0)) {
        // no room to write START_PACKET; drop the packet and try again next epoch.
        debugf(TRACE, "Transmit buffer is full.");
        return false;
    }

    // sent a START_PACKET, so increment pkts_sent.
    exc->pkts_sent++;

    size_t actually_written = fakewire_enc_encode_data(&exc->encoder, exc->conf->write_buffer, length);
    if (actually_written < length) {
        // not enough room to finish writing the whole packet at once; drop it.
        exc->write_needs_error = true;
        debugf(TRACE, "Transmit buffer is either full or not large enough.");
        return false;
    }

    if (!fakewire_enc_encode_ctrl(&exc->encoder, FWC_END_PACKET, 0)) {
        // no room to write END_PACKET; drop it. (disappointing, but the alternative is transmitting a packet late.)
        exc->write_needs_error = true;
        debugf(TRACE, "Transmit buffer is full.");
        return false;
    }

#ifdef EXCHANGE_DEBUG
    debug_printf(TRACE, "Transmitted packet (len=%u).", length);
#endif

    return true;
}

void fakewire_exc_exchange_loop(const fw_exchange_t *conf) {
    assert(conf != NULL);
    exchange_instance_t *exc = conf->instance;
    assert(exc != NULL);

    memset(exc, 0, sizeof(*exc));
    exchange_instance_configure(exc, conf, handshake_period_ticks());
    fakewire_enc_init(&exc->encoder, conf->transmit_chart);
    fakewire_dec_init(&exc->decoder, conf->receive_chart);

    debug_printf(DEBUG, "First handshake scheduled for %u ticks in the future", exc->countdown_timeout);

    while (true) {
        exchange_instance_check_invariants(exc);

        duct_receive_prepare(exc->conf->write_duct, EXCHANGE_REPLICA_ID);

        duct_flow_index dropped = 0;
        size_t packet_length;
        assert(exc->conf->buffers_length == duct_message_size(exc->conf->write_duct));
        while (0 != (packet_length = duct_receive_message(exc->conf->write_duct, EXCHANGE_REPLICA_ID,
                                                          exc->conf->write_buffer, NULL))) {
            assert(0 < packet_length && packet_length <= exc->conf->buffers_length);
            if (!exchange_instance_transmit_data(exc, packet_length)) {
                dropped++;
            }
        }
        if (dropped) {
            debug_printf(WARNING, "Dropped %u packets blocked from transmission.", dropped);
        }

        duct_receive_commit(exc->conf->write_duct, EXCHANGE_REPLICA_ID);

        // flush encoder before we sleep
        fakewire_enc_flush(&exc->encoder);

        // wait until we're scheduled again
        task_yield();

        exchange_instance_check_timers(exc);

        duct_send_prepare(exc->conf->read_duct, EXCHANGE_REPLICA_ID);
        // keep receiving line data as long as there's more data to receive; we don't want to sleep until there's
        // nothing left, so that we can guarantee a wakeup will still be pending afterwards,
        while (exchange_instance_receive(exc)) {
            // keep looping
        }
        duct_send_commit(exc->conf->read_duct, EXCHANGE_REPLICA_ID);

        exchange_instance_check_fcts(exc);

        exchange_instance_transmit_tokens(exc);

        exchange_instance_transmit_handshakes(exc);

        // wait until we're scheduled again before we try to receive
        task_yield();
    }
}

File: sim/fakewire/exchange/exchange.go
package exchange

import (
	"fmt"
	"github.com/celskeggs/hailburst/sim/component"
	"github.com/celskeggs/hailburst/sim/fakewire/codec"
	"github.com/celskeggs/hailburst/sim/fakewire/fwmodel"
	"github.com/celskeggs/hailburst/sim/model"
	"log"
	"time"
)

type ExchangeState uint8

const (
	MaxOutstandingTokens = 10
)

const DetailedDebug = false

const (
	StateInvalid = iota
	StateConnecting
	StateHandshaking
	StateOperating
)

func (s ExchangeState) String() string {
	switch s {
	case StateConnecting:
		return "CONNECTING"
	case StateHandshaking:
		return "HANDSHAKING"
	case StateOperating:
		return "OPERATING"
	default:
		return fmt.Sprintf("[UNKNOWN=%d]", uint8(s))
	}
}

type Exchange struct {
	Sim   model.SimContext
	Label string

	State ExchangeState

	SendHandshakeId        uint32
	RecvHandshakeId        uint32
	SendSecondaryHandshake bool

	FctsSent uint32
	FctsRcvd uint32
	PktsSent uint32
	PktsRcvd uint32

	InboundBuffer  []byte
	InboundPending [][]byte
	RecvInProgress bool

	CondNotify *component.EventDispatcher

	TxBusy bool
}

func (ex *Exchange) CheckInvariants() {
	if ex.State < StateConnecting || ex.State > StateOperating {
		log.Panicf("invalid state: %v", ex.State)
	}
	if ex.FctsRcvd < ex.PktsSent || ex.FctsRcvd > ex.PktsSent+MaxOutstandingTokens {
		log.Panicf("invalid pkts_sent=%v fcts_rcvd=%v", ex.PktsSent, ex.FctsRcvd)
	}
}

func (ex *Exchange) Reset() {
	ex.State = StateConnecting

	ex.SendHandshakeId = 0
	ex.RecvHandshakeId = 0
	ex.SendSecondaryHandshake = false

	ex.FctsSent = 0
	ex.FctsRcvd = 0
	ex.PktsSent = 0
	ex.PktsRcvd = 0

	ex.InboundBuffer = nil
	ex.InboundPending = nil
	ex.RecvInProgress = false

	ex.CheckInvariants()

	ex.CondNotify.DispatchLater()
}

func (ex *Exchange) Debug(explanation string, args ...interface{}) {
	log.Printf("%v [%s] FakeWire: %s", ex.Sim.Now(), ex.Label, fmt.Sprintf(explanation, args...))
}

func (ex *Exchange) ReceiveBytes(bytes []byte) {
	ex.CheckInvariants()

	if ex.State == StateOperating {
		// receive packet bytes during operating mode
		if !ex.RecvInProgress {
			ex.Debug("Hit unexpected data character 0x%x before start-of-packet; resetting.", bytes[0])
			ex.Reset()
		} else {
			ex.InboundBuffer = append(ex.InboundBuffer, bytes...)
		}
	} else {
		ex.Debug("Received unexpected data character 0x%x during handshake mode %v; resetting.", bytes[0], ex.State)
		ex.Reset()
	}
}

func (ex *Exchange) ReceiveCtrlChar(symbol codec.ControlChar, param uint32) {
	ex.CheckInvariants()
	symbol.Validate(param)
	if ex.State == StateConnecting {
		switch symbol {
		case codec.ChHandshake1:
			// received primary handshake
			ex.Debug("Received a primary handshake with ID=0x%08x.", param)
			ex.RecvHandshakeId = param
			ex.SendSecondaryHandshake = true
			ex.CondNotify.DispatchLater()
		case codec.ChHandshake2:
			ex.Debug("Received unexpected secondary handshake when no primary handshake had been sent; resetting.")
			ex.Reset()
		default:
			ex.Debug("Hit unexpected control character %v while CONNECTING; resetting.", symbol)
			ex.Reset()
		}
	} else if ex.State == StateHandshaking {
		switch symbol {
		case codec.ChHandshake1:
			ex.Debug("Received primary handshake collision while handshaking; resetting.")
			ex.Reset()
		case codec.ChHandshake2:
			// received secondary handshake
			if param == ex.SendHandshakeId {
				ex.Debug("Received secondary handshake with ID=0x%08x; transitioning to operating mode.", param)
				ex.State = StateOperating
				ex.CondNotify.DispatchLater()
			} else {
				ex.Debug("Received mismatched secondary ID 0x%08x instead of 0x%08x; resetting.", param, ex.SendHandshakeId)
				ex.Reset()
			}
		default:
			ex.Debug("Hit unexpected control character %v while HANDSHAKING; resetting.", symbol)
			ex.Reset()
		}
	} else if ex.State == StateOperating {
		switch symbol {
		case codec.ChHandshake1:
			// abort connection and restart everything; handle primary handshake
			ex.Debug("Received a primary handshake with ID=0x%08x during operating mode; resetting.", param)
			ex.Reset()
			ex.RecvHandshakeId = param
			ex.SendSecondaryHandshake = true
			ex.CondNotify.DispatchLater()
		case codec.ChHandshake2:
			// abort connection and restart everything
			ex.Debug("Received secondary handshake request during operating mode; resetting.")
			ex.Reset()
		case codec.ChStartPacket:
			if ex.FctsSent <= ex.PktsRcvd {
				ex.Debug("Received unauthorized start-of-packet (fcts_sent=%v, pkts_rcvd=%v); resetting.",
					ex.FctsSent, ex.PktsRcvd)
				ex.Reset()
			} else {
				if ex.RecvInProgress || len(ex.InboundBuffer) > 0 {
					panic("inconsistent state with HasSentFCT")
				}
				ex.InboundBuffer = []byte{}
				ex.RecvInProgress = true
				ex.PktsRcvd += 1
			}
		case codec.ChEndPacket:
			if !ex.RecvInProgress {
				ex.Debug("Hit unexpected end-of-packet before start-of-packet; resetting.")
				ex.Reset()
			} else {
				if ex.PktsRcvd > ex.FctsSent {
					log.Panicf("inconsistent state:\n\tRecvInProgress=%v\n\tPktsRcvd=%v\n\tFctsSent=%v\n",
						ex.RecvInProgress, ex.PktsRcvd, ex.FctsSent)
				}
				ex.InboundPending = append(ex.InboundPending, ex.InboundBuffer)
				ex.InboundBuffer = nil
				ex.RecvInProgress = false
				ex.CondNotify.DispatchLater()
			}
		case codec.ChErrorPacket:
			if !ex.RecvInProgress {
				ex.Debug("Hit unexpected error-end-of-packet before start-of-packet; resetting.")
				ex.Reset()
			} else {
				if ex.PktsRcvd > ex.FctsSent {
					log.Panicf("inconsistent state:\n\tRecvInProgress=%v\n\tPktsRcvd=%v\n\tFctsSent=%v\n",
						ex.RecvInProgress, ex.PktsRcvd, ex.FctsSent)
				}
				// discard the data in the current packet
				ex.InboundBuffer = nil
				ex.RecvInProgress = false
			}
		case codec.ChFlowControl:
			if param < ex.FctsRcvd {
				// FCT number should never decrease
				ex.Debug("Received abnormally low FCT(%v) when last count was %v; resetting.", param, ex.FctsRcvd)
				ex.Reset()
			} else if param > ex.PktsSent+MaxOutstandingTokens {
				// FCT number should never increase more than allowed.
				ex.Debug("Received abnormally high FCT(%v) when maximum was %d and last count was %d; resetting.",
					param, ex.PktsSent+MaxOutstandingTokens, ex.FctsRcvd)
				ex.Reset()
			} else {
				// received FCT; may be able to send more packets!
				ex.FctsRcvd = param
				ex.CondNotify.DispatchLater()
			}
		case codec.ChKeepAlive:
			if ex.PktsRcvd != param {
				ex.Debug("KAT mismatch: received %v packets, but supposed to have received %v; resetting.",
					ex.PktsRcvd, param)
				ex.Reset()
			}
		case codec.ChEscapeSym:
			// indicates that an invalid escape sequence was received by the codec
			ex.Debug("Received invalid escape sequence; resetting.")
			ex.Reset()
		default:
			log.Panicf("invalid control character: %v", symbol)
		}
	} else {
		log.Panicf("Invalid state: %v", ex.State)
	}
}

// Transmit requires that TxBusy is false.
func (ex *Exchange) Transmit(io *component.TwixtIO, sink model.DataSinkBytes, data []byte) {
	if ex.TxBusy {
		panic("Transmit contract requires TxBusy=false on entry")
	}
	ex.TxBusy = true
	for len(data) > 0 {
		count := sink.TryWrite(data)
		data = data[count:]
		if len(data) > 0 {
			io.Yield()
			ex.CheckInvariants()
		}
	}
	if !ex.TxBusy {
		panic("inconsistent: TxBusy should still be set")
	}
	ex.TxBusy = false
	ex.CondNotify.DispatchLater()
}

func concat(allData ...[]byte) (total []byte) {
	for _, data := range allData {
		if len(total) == 0 {
			total = data
		} else if len(data) > 0 {
			total = append(total, data...)
		}
	}
	return total
}

// HandshakePeriod generates a random interval in the range [3ms, 10ms)
func (ex *Exchange) HandshakePeriod() time.Duration {
	msInNs := time.Millisecond.Nanoseconds()
	return time.Duration(ex.Sim.Rand().Int63n(7*msInNs)+3*msInNs) * time.Nanosecond
}

func FakeWire(ctx model.SimContext, lsink model.DataSinkBytes, lsource model.DataSourceBytes, psink fwmodel.PacketSink, psource fwmodel.PacketSource, label string) {
	// data receiver thread

	ex := &Exchange{
		Sim:        ctx,
		Label:      label,
		CondNotify: component.MakeEventDispatcher(ctx, "sim.fakewire.exchange.FakeWire/CondNotify"),
	}
	ex.Reset()

	// connect a Decoder to receive all data from the link source
	decoder := codec.MakeDecoder(ex.ReceiveBytes, ex.ReceiveCtrlChar)
	component.DataPumpDirect(ctx, lsource, decoder.Decode)

	// Receiver Thread
	component.BuildTwixt(ctx, []model.EventSource{psink, ex.CondNotify}, func(io *component.TwixtIO) {
		ex.CheckInvariants()
		for {
			pending := uint32(len(ex.InboundPending))
			if ex.RecvInProgress {
				pending += 1
			}
			if ex.State == StateOperating && !ex.TxBusy && ex.FctsSent < ex.PktsRcvd+MaxOutstandingTokens-pending {
				ex.FctsSent += 1
				ex.Transmit(io, lsink, codec.EncodeCtrlChar(codec.ChFlowControl, ex.FctsSent))
			} else if len(ex.InboundPending) > 0 && psink.CanAcceptPacket() {
				psink.SendPacket(ex.InboundPending[0])
				ex.InboundPending = ex.InboundPending[1:]
			} else {
				// if nothing to do, yield
				io.Yield()
				ex.CheckInvariants()
			}
		}
	})
	// Transmitter thread
	component.BuildTwixt(ctx, []model.EventSource{psource, lsink, ex.CondNotify}, func(io *component.TwixtIO) {
		ex.CheckInvariants()
		for {
			// wait until handshake completes, transmit is possible, and a packet is ready to be sent
			for ex.State != StateOperating || ex.PktsSent >= ex.FctsRcvd || ex.TxBusy || !psource.HasPacketAvailable() {
				io.Yield()
				ex.CheckInvariants()
			}
			packet := psource.ReceivePacket()
			ex.PktsSent += 1
			ex.Transmit(io, lsink, concat(
				codec.EncodeCtrlChar(codec.ChStartPacket, 0),
				codec.EncodeDataBytes(packet),
				codec.EncodeCtrlChar(codec.ChEndPacket, 0),
			))
		}
	})
	// Flow token thread
	component.BuildTwixt(ctx, []model.EventSource{lsink, ex.CondNotify}, func(io *component.TwixtIO) {
		ex.CheckInvariants()
		nextInterval := ctx.Now().Add(ex.HandshakePeriod())
		if DetailedDebug {
			ex.Debug("First handshake scheduled for: %v", nextInterval)
		}
		// requires that TxBusy is false
		sendCtrlChar := func(ctrlChar codec.ControlChar, param uint32) {
			ex.Transmit(io, lsink, concat(
				codec.EncodeCtrlChar(ctrlChar, param),
			))
			nextInterval = ctx.Now().Add(ex.HandshakePeriod())
			if DetailedDebug {
				ex.Debug("Next handshake scheduled for: %v", nextInterval)
			}
		}
		for {
			for ex.TxBusy {
				io.Yield()
				ex.CheckInvariants()
			}
			if ex.SendSecondaryHandshake {
				if ex.State != StateConnecting {
					panic("invalid state")
				}
				// if indicated, send secondary handshake
				handshakeId := ex.RecvHandshakeId

				sendCtrlChar(codec.ChHandshake2, handshakeId)

				if !ex.SendSecondaryHandshake {
					ex.Debug("Sent secondary handshake with ID=0x%08x, but request revoked by reset; not transitioning.", handshakeId)
				} else if handshakeId != ex.RecvHandshakeId {
					ex.Debug("Sent secondary handshake with ID=0x%08x, but new primary ID=0x%08x had been received in the meantime; not transitioning.",
						handshakeId, ex.RecvHandshakeId)
				} else if ex.State != StateConnecting {
					ex.Debug("Sent secondary handshake with ID=0x%08x, but state is now %v instead of CONNECTING; not transitioning.",
						handshakeId, ex.State)
				} else {
					ex.Debug("Sent secondary handshake with ID=0x%08x; transitioning to operating mode.", handshakeId)
					ex.State = StateOperating
					ex.SendSecondaryHandshake = false
				}

				ex.CondNotify.DispatchLater()
			} else if ex.State == StateOperating {
				// if we're operating... then we need to send FCTs and KATs on a regular basis
				if ctx.Now().Before(nextInterval) {
					io.YieldUntil(nextInterval)
					ex.CheckInvariants()
					continue
				}
				sendCtrlChar(codec.ChFlowControl, ex.FctsSent)
				sendCtrlChar(codec.ChKeepAlive, ex.PktsSent)
			} else {
				// if we're in the handshaking state... then we need to send primary handshakes on a regular basis
				if ctx.Now().Before(nextInterval) {
					io.YieldUntil(nextInterval)
					ex.CheckInvariants()
					continue
				}
				// pick something very likely to be distinct (Go picks msb unset, C picks msb set)
				handshakeId := 0x7FFFFFFF & ctx.Rand().Uint32()
				ex.SendHandshakeId = handshakeId
				ex.State = StateHandshaking

				sendCtrlChar(codec.ChHandshake1, handshakeId)

				ex.Debug("Sent primary handshake with ID=0x%08x.", handshakeId)
			}
		}
	})
}
